napari/_qt/layer_controls/qt_shapes_controls.py

```diff
@@ -410,3 +410,10 @@ def _on_editable_change(self, event=None):
             ],
             self.layer.editable,
         )
+
+    def close(self):
+        """Layer widget is closing."""
+        self.layer._text.events.visible.disconnect(
+            self._on_text_visibility_change
+        )
+        super().close()
```

Comment:
The issue here is that the connections are happening to `self.layer._text` not `self.layer`. Not all layers have a `_text`. I could check if they do and disconnect everything from it in the base, but otherwise this will require some special handling. Similar for the connection in the first place. Now I'd need a list for `text_connections`. This might also be replicated when we add a ColorManager, see #1940

Reply:
I hope this is more clear now